django-stubs/db/models/expressions.pyi

```diff
@@ -199,8 +202,14 @@ class ExpressionList(Func):
 
 class OrderByList(Func): ...
 
-class ExpressionWrapper(Expression):
-    def __init__(self, expression: Q | Combinable, output_field: Field) -> None: ...
+_E = TypeVar("_E", bound=Q | Combinable)
+
+class ExpressionWrapper(Expression, Generic[_E]):
+    def __init__(self, expression: _E, output_field: Field) -> None: ...
+    expression: _E
+
+class NegatedExpression(ExpressionWrapper):
```

Comment:
Shouldn't it also be generic? Since its parent is generic.

Reply:
Locally, I also made the following changes: 

```diff
diff --git a/django-stubs/db/models/expressions.pyi b/django-stubs/db/models/expressions.pyi
index 823b436b..296f0b3c 100644
--- a/django-stubs/db/models/expressions.pyi
+++ b/django-stubs/db/models/expressions.pyi
@@ -1,7 +1,7 @@
 import datetime
 from collections.abc import Callable, Iterable, Iterator, Sequence
 from decimal import Decimal
-from typing import Any, Generic, Literal, TypeVar
+from typing import Any, Generic, Literal, TypeVar, overload
 
 from _typeshed import Self
 from django.db.backends.base.base import BaseDatabaseWrapper
@@ -54,7 +54,7 @@ class Combinable:
     def __rand__(self, other: Any) -> Combinable: ...
     def __ror__(self, other: Any) -> Combinable: ...
     def __rxor__(self, other: Any) -> Combinable: ...
-    def __invert__(self) -> Combinable: ...
+    def __invert__(self: Self) -> NegatedExpression[Self]: ...
 
 class BaseExpression:
     is_summary: bool
@@ -208,8 +208,12 @@ class ExpressionWrapper(Expression, Generic[_E]):
     def __init__(self, expression: _E, output_field: Field) -> None: ...
     expression: _E
 
-class NegatedExpression(ExpressionWrapper):
-    def __init__(self, expression: Q | Combinable) -> None: ...
+class NegatedExpression(ExpressionWrapper, Generic[_E]):
+    def __init__(self, expression: _E) -> None: ...
+    @overload
+    def __invert__(self: NegatedExpression[Q]) -> Q: ...
+    @overload
+    def __invert__(self: NegatedExpression[Combinable]) -> Combinable: ...
 
 class When(Expression):
     template: str
```

but then, I get:
```bash
$ mypy --cache-dir=/dev/null --no-incremental django-stubs
django-stubs/db/models/expressions.pyi:57: error: Type argument "Self" of "NegatedExpression" must be a subtype of "Union[Q, Combinable]"
django-stubs/db/models/expressions.pyi:213: error: Signature of "__invert__" incompatible with supertype "Combinable"
django-stubs/db/models/expressions.pyi:213: note:      Superclass:
django-stubs/db/models/expressions.pyi:213: note:          def __invert__(self) -> NegatedExpression[NegatedExpression[_E]]
django-stubs/db/models/expressions.pyi:213: note:      Subclass:
django-stubs/db/models/expressions.pyi:213: note:          @overload
django-stubs/db/models/expressions.pyi:213: note:          def __invert__(self) -> Q
django-stubs/db/models/expressions.pyi:213: note:          @overload
django-stubs/db/models/expressions.pyi:213: note:          def __invert__(self) -> Combinable
django-stubs/contrib/postgres/search.pyi:62: error: Return type "SearchQuery" of "__invert__" incompatible with return type "NegatedExpression[SearchQuery]" in supertype "Combinable"
Found 3 errors in 2 files (checked 683 source files)
```
and couldn’t find a way to solve it.

Reply:
You can always add `type: ignore[override]`

Reply:
Maybe just this:
```diff
diff --git a/django-stubs/db/models/expressions.pyi b/django-stubs/db/models/expressions.pyi
index 823b436b..60748ee6 100644
--- a/django-stubs/db/models/expressions.pyi
+++ b/django-stubs/db/models/expressions.pyi
@@ -1,7 +1,7 @@
 import datetime
 from collections.abc import Callable, Iterable, Iterator, Sequence
 from decimal import Decimal
-from typing import Any, Generic, Literal, TypeVar
+from typing import Any, Generic, Literal, TypeVar, overload
 
 from _typeshed import Self
 from django.db.backends.base.base import BaseDatabaseWrapper
@@ -208,8 +208,12 @@ class ExpressionWrapper(Expression, Generic[_E]):
     def __init__(self, expression: _E, output_field: Field) -> None: ...
     expression: _E
 
-class NegatedExpression(ExpressionWrapper):
-    def __init__(self, expression: Q | Combinable) -> None: ...
+class NegatedExpression(ExpressionWrapper, Generic[_E]):
+    def __init__(self, expression: _E) -> None: ...
+    @overload
+    def __invert__(self: NegatedExpression[Q]) -> Q: ...
+    @overload
+    def __invert__(self: NegatedExpression[Combinable]) -> Combinable: ...
 
 class When(Expression):
     template: str
```
and then `mypy` is ok. But then I feel we loose the relation between `Combinable` and `NegatedExpression` and it feels a bit sad :shrug:

Reply:
I haven't looked into this discussion in detail. But I've found that `typing_extensions.Self` sometimes fixes issues that occur when using `_typeshed.Self`, maybe something to try.

Reply:
Disregard what I wrote above :)

I looked into this and I agree, `# type: ignore` seems the best option here. It's technically a violation of LSP, but I think it's more useful than keeping `Combinable.__invert__() -> Combinable`.

Reply:
Any drawbacks to typehinting this as `def __invert__(self) -> _E: ...`?

> `def __invert__(self: NegatedExpression[Q]) -> Q: ...`

Can this actually occur?

ISTM `NegatedExpression` is not constructed with typevar _E=Q. I think _E can only be `Combinable`.

Because `Q.__invert__()` returns `Q`, not `NegatedExpression[Q]`

Reply:
I think you’re right, `def __invert__(self) -> _E: ...` is good enough, I’m not sure why I’m complicating things. I’m pushing this and will do a self review.